From e82abab192345a57df5877486d1a30599632eba8 Mon Sep 17 00:00:00 2001
From: Jeroen Hofstee <jeroen@myspectrum.nl>
Date: Sun, 24 Nov 2019 17:02:08 +0100
Subject: [PATCH] fw_env: use default environment if open failed

This brings the behaviour for libuboot in sync with what
u-boot itself does. This seems fine for a rpi, must be
better looked at for general things.

Upstream-Status: Inappropriate [Venus specific]

Signed-off-by: Leon Anavi <leon.anavi@konsulko.com>
---
 tools/env/fw_env.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/tools/env/fw_env.c b/tools/env/fw_env.c
index 425faf380fb..98467c151ad 100644
--- a/tools/env/fw_env.c
+++ b/tools/env/fw_env.c
@@ -520,7 +520,7 @@ int fw_env_flush(struct env_opts *opts)
 	*environment.crc = crc32(0, (uint8_t *) environment.data, ENV_SIZE);
 
 	/* write environment back to flash */
-	if (flash_io(O_RDWR, environment.image, CUR_ENVSIZE)) {
+	if (flash_io(O_RDWR | O_CREAT, environment.image, CUR_ENVSIZE)) {
 		fprintf(stderr, "Error: can't write fw_env to flash\n");
 		return -1;
 	}
@@ -1390,7 +1390,7 @@ static int flash_io(int mode, void *buf, size_t count)
 		return -1;
 	}
 
-	if (mode == O_RDWR) {
+	if (mode & O_RDWR) {
 		rc = flash_io_write(fd_current, buf, count);
 	} else {
 		rc = flash_read(fd_current, buf, count);
@@ -1480,8 +1480,10 @@ int fw_env_open(struct env_opts *opts)
 		redundant1 = buf1;
 
 		if (flash_io(O_RDONLY, buf1, CUR_ENVSIZE)) {
-			ret = -EIO;
-			goto open_cleanup;
+			fprintf(stderr, "Warning: Open failed, using default environment\n");
+			memcpy(environment.data, default_environment,
+ï¿¼				sizeof(default_environment));
+			dev_current = 0;
 		}
 
 		/* Check flag scheme compatibility */
@@ -1619,7 +1621,8 @@ static int check_device_config(int dev)
 	if (fd < 0) {
 		fprintf(stderr,
 			"Cannot open %s: %s\n", DEVNAME(dev), strerror(errno));
-		return -1;
+		// Fine in case of a file, it will be created.
+		return 0;
 	}
 
 	rc = fstat(fd, &st);
-- 
2.44.3

